### PR TITLE
Implement an API Redirect registration system (spike)

### PR DESCRIPTION
**Pull request by sgmiller**

### CODE REVIEW

File: http/handler.go [hunk 1]
```diff
@@ -412,9 +412,18 @@ func wrapGenericHandler(core *vault.Core, h http.Handler, props *vault.HandlerPr
...
 		case strings.HasPrefix(r.URL.Path, "/ui"), r.URL.Path == "/robots.txt", r.URL.Path == "/":
 		default:
+			redir, err := core.GetAPIRedirect(r.Context(), r.URL.Path)
+			if err != nil {
+				core.Logger().Warn("error resolving potential API redirect", "error", err)
+			} else {
+				if redir != "" {
+					w.Header().Set("Location", redir)
+					w.WriteHeader(http.StatusFound)
+					return
```
Comment (stevendpclark) [inline http/handler.go:424]: Missing cancelFunc() call?

File: http/api_redirect.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+func handleApiRedirect(core *vault.Core, base http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		redir, err := core.GetAPIRedirect(r.Context(), r.URL.Path)
+		if err != nil {
+			core.Logger().Warn("error resolving potential API redirect", "error", err)
+		} else {
+			if redir != "" {
+				w.Header().Set("Location", redir)
+				w.WriteHeader(http.StatusFound)
```
Comment (stevendpclark) [inline http/api_redirect.go:17]: Should we make this logic redirect 302 on GET/HEAD requests only and use 307 for other request types?

I've been burned by bad browser behavior [in the past around 302 handling](https://en.wikipedia.org/wiki/HTTP_302)...

File: sdk/logical/logical.go [hunk 3]
```diff
@@ -159,6 +159,9 @@ type Paths struct {
 	// On standby nodes, like all storage write operations, this will trigger
 	// an ErrReadOnly return.
 	WriteForwardedStorage []string
+
+	// RFC5785
+	WellKnown map[string]string
```
Comment (stevendpclark) [inline sdk/logical/logical.go:164]: unused? Also I'm not seeing the hook for disabling a mount removing it's entries from the redirect map?

File: http/api_redirect.go (hunk 2 above)
Comment (sgmiller) [reply http/api_redirect.go:17]: TiL

**Comment by schavis** [comment]
I don't see any documentation in the changed files, so I'm removing the vault-education-approvers reviewer. If you decide to add doc changes to the spike, please make sure to add us back for review :)

**Pull request closed without merging**